module/x/gravity/keeper/ethereum_event_handler.go

```diff
@@ -121,3 +78,102 @@ func (a EthereumEventProcessor) Handle(ctx sdk.Context, eve types.EthereumEvent)
+		)
+	}
+
+	// We expect that all Cosmos-based tokens have metadata defined. In the case
+	// a token does not have metadata defined, e.g. an IBC token, we successfully
+	// handle the token under the following conditions:
+	//
+	// 1. The ERC20 name is equal to the token's denomination. Otherwise, this
+	// 		means that ERC20 tokens would have an untenable UX.
+	// 2. The ERC20 token has zero decimals as this is what we default to since
+	// 		we cannot know or infer the real decimal value for the Cosmos token.
+	// 3. The ERC20 symbol is empty.
+	//
+	// NOTE: This path is not encouraged and all supported assets should have
+	// metadata defined. If metadata cannot be defined, consider adding the token's
+	// metadata on the fly.
+	metadata := a.keeper.bankKeeper.GetDenomMetaData(ctx, event.CosmosDenom)
+	if metadata.Base == "" {
+		if event.Erc20Name != event.CosmosDenom {
+			return sdkerrors.Wrapf(
+				types.ErrInvalid,
+				"invalid ERC20 name for token without metadata; got: %s, expected: %s", event.Erc20Name, event.CosmosDenom,
+			)
+		}
+
+		if event.Erc20Symbol != "" {
+			return sdkerrors.Wrapf(
+				types.ErrInvalid,
+				"expected empty ERC20 symbol for token without metadata; got: %s", event.Erc20Symbol,
+			)
+		}
+
+		if event.Erc20Decimals != 0 {
+			return sdkerrors.Wrapf(
+				types.ErrInvalid,
+				"expected zero ERC20 decimals for token without metadata; got: %d", event.Erc20Decimals,
+			)
+		}
+
+		return nil
+	}
+
+	return verifyERC20Token(metadata, event)
+}
+
+func verifyERC20Token(metadata banktypes.Metadata, event *types.ERC20DeployedEvent) error {
+	if event.Erc20Name != metadata.Display {
+		return sdkerrors.Wrapf(
+			types.ErrInvalid,
+			"ERC20 name %s does not match the denom display %s", event.Erc20Name, metadata.Description,
+		)
+	}
+
+	if event.Erc20Symbol != metadata.Display {
+		return sdkerrors.Wrapf(
+			types.ErrInvalid,
+			"ERC20 symbol %s does not match denom display %s", event.Erc20Symbol, metadata.Display,
+		)
+	}
+
+	// ERC20 tokens use a very simple mechanism to tell you where to display the
+	// decimal point. The "decimals" field simply tells you how many decimal places
+	// there will be.
+	//
+	// Cosmos denoms have a system that is much more full featured, with
+	// enterprise-ready token denominations. There is a DenomUnits array that
+	// tells you what the name of each denomination of the token is.
+	//
+	// To correlate this with an ERC20 "decimals" field, we have to search through
+	// the DenomUnits array to find the DenomUnit which matches up to the main
+	// token "display" value. Then we take the "exponent" from this DenomUnit.
+	//
+	// If the correct DenomUnit is not found, it will default to 0. This will
+	// result in there being no decimal places in the token's ERC20 on Ethereum.
+	// For example, if this happened with ATOM, 1 ATOM would appear on Ethereum
+	// as 1 million ATOM, having 6 extra places before the decimal point.
+	var decimals uint32
```

Comment:
ditto, I think we can use `sdk.Precision` which is 6

Reply:
See comment above. I don't think that's safe.